cloud provider refactor

ibgreen opened the pull request:
(This is a prep diff for an upcoming `GoogleDriveProvider` PR)

Lightly refactors, simplifies and fixes the cloud provider architecture to support additional providers.

Refactor:
- The providers are now class instances (instead of global functions) 
- The `DropboxProvider` is refactored as a class 

Reasons:
- Keeps client ids and names separate from implementation (see `src/cloud-providers/providers.js`)
- Allows for general code to be written for a certain authentication system/SDK.

Simplifications: 
- Cloud Providers moved to top-level directory: `src/cloud-providers` (since this is an important integration point it should be easy to find).
- The provider list is moved to `src/cloud-providers/providers.js` and is now an array (instead of a map - providers already have a name, no need to list it twice).

Bug fixes:
- Plumbs through the `providerName` in a couple of places where the code was always using the first provider in the list.

ibgreen commented:
@neokore I believe this contains a few of the fixes in your PR #830 (If this lands first, it will likely require you to do a small rebase, but overall this PR should improves the base you are building on).

heshan0131 commented:
> @ibgreen There is an issue when the user tries to log on dropbox. The new tab where Kepler receives the oatu2 callback has some errors with blank page and no dropbox token stored into local storage. I created a gif to reproduce the issue
> 
> ![dropbox-report](https://user-images.githubusercontent.com/2554552/69001360-b0571a80-0891-11ea-9f90-efe470e8f768.gif)

@macrigiuseppe  Ib doesn't have our dropboxcliendid to test the log in behavior. Could you help debug this error?

ibgreen commented on examples/demo-app/src/components/sharing/cloud-tile.js:
```diff
@@ -63,11 +63,12 @@ const TileButton = styled.button`
+const CloudTile = ({token, onExport, onLogin, Icon, name}) => {
   return (
-    <StyledTileWrapper onClick={token ? onExport : onLogin}>
+    <StyledTileWrapper onClick={token ? onExport.bind(null, name) : onLogin}>
```
Good idea. Added `useCallback`, please take another look to make sure I got it right.

ibgreen commented:
@macrigiuseppe 

Also as soon as this lands I will update the `GoogleDriveProvider` PR #832 (and yes, that one I have tested :)

To land that one, we will need a client id for Google Drive.
- One option is that I could provide a client id for Google Drive.
- The dropbox client is "hosted by" Uber but I assume that is not a requirement for other integrations.

Let me know (we should probably discuss in that PR...)

macrigiuseppe commented:
> > There is an issue when the user tries to log on dropbox. The new tab where Kepler receives the oatu2 callback has some errors with blank page and no dropbox token stored into local storage. I created a gif to reproduce the issue
> 
> @macrigiuseppe Thanks for the screenshot, that error was due to an incorrect import path. I have added a fix for that, but as @heshan0131 says I cannot do a full test of the dropbox integration so there may be some other issue, happy to fix if you can show me what is breaking.

@ibgreen you can simply create a new dropbox client id using any dropbox account. Here is a guide: https://auth0.com/docs/connections/social/dropbox

ibgreen commented:
> Let me know if you can test it

Yes, I suppose I should, though it could take some time before I get it done.

And this is just one provider... longer term it seems like this could grow into a rather difficult challenge for the maintenance of the cloud provider system.

The testing matrix will become more complicated as more cloud providers are integrated, especially as developers who want to contribute don't have access to a set of test accounts and client ids but have to start individually signing up for two different things (account + client id) per provider...

macrigiuseppe commented:
> > Let me know if you can test it
> 
> Yes, I suppose I should, though it could take some time before I get it done.
> 
> And this is just one provider... longer term it seems like this could grow into a rather difficult challenge for the maintenance of the cloud provider system.
> 
> The testing matrix will become more complicated as more cloud providers are integrated, especially as developers who want to contribute don't have access to a set of test accounts and client ids but have to start individually signing up for two different things (account + client id) per provider...

I think in the long term we should create some e2e testing case that could actually validate all these cases so every time we introduce a new provider we can add new tests.

I created issue #834 and assigned to myself.

macrigiuseppe commented on examples/demo-app/src/cloud-providers/dropbox-provider.js:
```diff
@@ -0,0 +1,164 @@
+          timestamp: new Date()
+        }));
+      }
+      onCloudLoginSuccess(NAME);
```
should this be
```
onCloudLoginSuccess(this.name);
```

ibgreen commented:
> Code looks good, just a few more comments

Addressed the `this.name` nit, and updated the PR.

> Tested flow and it works. 

Thanks!!!

I was actually able to test with the existing client id (it seems to be set up to allow localhost which is great 👍 ) However, it turns out that my dropbox account is full (since I stopped paying for it) so I get write failures. They are clearly reported as quota issues by dropbox, but this is not handled gracefully by current code (generates react errors).

I doubt this is introduced by my changes (I suspect the failure case was never well tested?). I might be able to debug it a bit more later, but in general, I think the cloud provider system first needs a mechanism for relaying errors back to the user. Right now there is little feedback on what actually happened during upload.

macrigiuseppe commented:
This is actually a good point.

Currently if you want to provide a visual feedback to users, eg errors, you
can use the notification system.

I will look into Dropbox integration to be more resilient to errors.



On Sun, Nov 17, 2019 at 9:20 AM Ib Green <notifications@github.com> wrote:

> Code looks good, just a few more comments
>
> Addressed the this.name nit, and updated the PR.
>
> Tested flow and it works.
>
> Thanks!!!
>
> I was actually able to test with the existing client id (it seems to be
> set up to allow localhost which is great 👍 ) However, it turns out that
> my dropbox account is full (since I stopped paying for it) so I get write
> failures. They are clearly reported as quota issues by dropbox, but this is
> not handled gracefully by current code (generates react errors).
>
> I doubt this is introduced by my changes (I suspect the failure case was
> never well tested?). I might be able to debug it a bit more later, but in
> general, I think the cloud provider system first needs a mechanism for
> relaying errors back to the user. Right now there is little feedback on
> what actually happened during upload.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keplergl/kepler.gl/pull/831?email_source=notifications&email_token=AATPVOHH2BQUHRDOIQQ4A6DQUF4OZA5CNFSM4JN7KII2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEIRFMI#issuecomment-554767025>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AATPVOGAHHROJQ26XKBRZGDQUF4OZANCNFSM4JN7KIIQ>
> .
>
-- 
Giuseppe Macrì

+1 415 463 9771
macri.giuseppe@gmail.com

VictorVelarde commented:
Nice refactor @ibgreen!, @neokore and I will work to adapt the current carto cloud provider (WIP) to this new base code.

Just a small comment on this:
https://github.com/keplergl/kepler.gl/blob/754b33421357d95c3e7ee162aa8e6050f0e0746d/examples/demo-app/src/components/sharing/cloud-storage.js#L165

It looks like this piece might be provider-specific.  Not all the providers would work with the `folderLink` concept (CARTO doesn't have one), and the copy would need to be customized too. What do you think?

ibgreen commented:
@VictorVelarde 

Yes I also think we are going to want to integrate cloud providers with more features than just save/load, so we will need a feature detection system. 

For now would it be sufficient if you just added a check for `null` folderlink, and changed Dropbox to something more general (or the cloud provider's name)?

Also, longer term, we need good documentation for the cloud provider architecture. E.g. the metadata object here should be documented so cloud providers know how to populate it, without having to discover this code. 

I can prepare an initial PR for such a doc.

VictorVelarde commented:
Thx! yes we can try with that approach by now, just checking for `null`. We will try to get those changes soon and update our open PR to your new design.

And we would be really interested in that 'extended' cloud provider architecture.
